drivers/clock_control/nrf_power_clock.c

```diff
@@ -459,18 +471,53 @@ static void lfclk_spinwait(nrf_clock_lfclk_t t)
 {
 	nrf_clock_domain_t d = NRF_CLOCK_DOMAIN_LFCLK;
 	nrf_clock_lfclk_t type;
+	static const bool lfxo_target =
+		    IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_XTAL) |
+		    IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_LOW_SWING) |
+		    IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING);
+	int key = irq_lock();
 
 	while (!(nrf_clock_is_running(NRF_CLOCK, d, (void *)&type)
 		 && (type == t))) {
-		/* empty */
-		/* k_cpu_idle(); */
+		/* Synth source start is almost instant and LFCLKSTARTED may
+		 * happen before calling idle. That would lead to deadlock.
+		 */
+		if (!IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_SYNTH)) {
```

Comment:
I see the goal here, but `z_clock_driver_init()` will invoke `z_nrf_clock_control_lf_on()` which on its first entry will use the onoff manager to trigger `TASK_LFCLKSTART`.  All that is before interrupts are locked on entry to this function.

There is a typical 1 ms startup time for this on nRF52840, but it does seem like the ISR may have already processed LFCLKSTARTED before the lock is taken, so deadlock is still possible.

Reply:
> There is a typical 1 ms startup time for this on nRF52840, but it does seem like the ISR may have already processed LFCLKSTARTED before the lock is taken, so deadlock is still possible.

Interrupts are locked at this point or startup. I call `irq_lock` just to get `key` which is required by `k_cpu_atomic_idle` and `irq_unlock` for consistency.

Reply:
i will see if i can do something to support call of `z_nrf_clock_control_lf_on` without that assumption.

Reply:
> Interrupts are locked at this point of startup.

Could you expand on this?

`z_clock_driver_init()` at PRE_KERNEL_2/CONFIG_SYSTEM_CLOCK_INIT_PRIORITY invokes
`z_nrf_clock_control_lf_on()` invokes
`onoff_request()` invokes
`lfclk_start()` triggers NRF_CLOCK_TASK_LFCLKSTART then
`z_nrf_clock_control_lf_on()` invokes `lfclk_spinwait()`

Nothing on that path appears to disables interrupts, so theoretically the `EVENT_LFCLKSTARTED` could have been cleared by `clock_event_check_and_clean()` in `nrf_power_clock_isr()` before the subsequent `lfclk_spinwait()` gets control and does disable interrupt, expecting to be able to see that event itself.

What am I missing?

Reply:
> Nothing on that path appears to disables interrupts

driver initialization is called from `z_cstart` which has interrupts disabled.

I have added support to spinwaiting from thread context and added test for that.